### DISCUSSION TITLE
grid search AutoAugment & RandAugment for detection tasks

### DISCUSSION THREAD

**Discussion by Hapemo** [ultralytics/ultralytics#19092] [Ideas]
I noticed that Ultralytics supports AutoAugment, RandAugment and AugMix for classification tasks but not detection tasks. Instead, detection tasks uses other augmentations like mosaic and mix up. Since RandAugment is fast and simple, may I ask why not use it for detection tasks too? 

On the paper I read for [RandAugment](https://arxiv.org/pdf/1909.13719v2), there are 2 parameters num_ops (N) and magnitude (M). I may be wrong, but I don't think the current implementation in ultralytics utilizes a search algorithm to find the optimal N and M number, instead, it's using default value provided by torchvision. With that, how relevant would it be to do a simple grid search across those 2 parameters, especially for detection tasks?

I'm doing a final year capstone project regarding this topic, so to stretch innovation a bit further, how feasible would it be to apply this grid search on a custom RandAugment algorithm using augmentations from albumentation instead of torchvision (since torchvision only uses a fixed set of 14 augmentation operations)?

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Hapemo, thank you for your interest in Ultralytics 🚀 and for bringing up such an interesting topic! 

We recommend exploring our [Documentation](https://docs.ultralytics.com/) for additional insights into augmentation strategies for both classification and detection tasks. While AutoAugment, RandAugment, and AugMix are currently optimized for classification tasks, detection-specific augmentations like mosaic and mixup are tailored for object detection's unique spatial and contextual needs. That said, your inquiry into adapting RandAugment or exploring an augmentation grid search for detection tasks is fascinating and could open up valuable new research directions!

If you're reporting a 🐛 **Bug** or missing functionality, we would appreciate a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) that demonstrates the behavior so we can understand it better and assist accordingly. 

For your custom training ❓ **Question**, here are a few things that would help us provide tailored insights:  
1. Details of your specific setup (e.g., model architecture, dataset)  
2. Examples or logs showing your attempts with detection augmentations  
3. If you've made modifications or implemented custom augmentations via Albumentations, a brief summary or code snippet would be great.  

We also encourage you to stay connected with the Ultralytics community for collaborative discussions! Join real-time conversations on [Discord](https://discord.com/invite/ultralytics) 🎧, participate in thoughtful discussions on [Discourse](https://community.ultralytics.com/), or share your findings on our [Subreddit](https://reddit.com/r/Ultralytics). You might come across like-minded individuals working on similar projects.  

## Upgrade

If you're integrating RandAugment-related features or exploring custom augmentations, make sure to use the latest `ultralytics` package to access the most up-to-date features and fixes. You can upgrade with the command below:

```bash
pip install -U ultralytics
```

## Environments

For development, ensure you're using a verified environment. YOLO models can run in various setups, including Jupyter Notebooks, Google Cloud, Amazon AWS, and Docker. Here's an example for running in hosted notebook environments:  
<a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>  

## Status

To verify the broader framework's stability, you can always check the [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule).  
<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>  

This automated response is here to assist you 😊, but an Ultralytics engineer will also review your question soon to provide additional guidance and support.

**Comment by glenn-jocher** [comment]
@Hapemo the current detection pipeline prioritizes augmentations like mosaic and mixup, which have shown stronger performance for detection, but experimenting with a custom RandAugment grid search on Albumentations could offer interesting insights for your project.

**Reply by Hapemo** [reply]
Okay thank you!

**Reply by glenn-jocher** [reply]
@Hapemo you're welcome! 😊 Feel free to reach out if you have any more questions as you work on your project. For continued support, consider exploring our documentation at https://docs.ultralytics.com and engaging with the YOLO community.

**Comment by Aaronuo** [comment]
@Hapemo  
Hello, my current project also needs to use augmentation in detection, I feel your idea is very good, I wonder if you have realized it? Can you share the code if it is already implemented?

**Reply by Hapemo** [reply]
Hi Aaronuo, yup I have created a system that has integrated rand augment into yolo detection tasks and also enabled disabling auto augmentation on ultralytics’s part. 

The augmentation operations i used is from albumentation, different fron pytorch’s implementation of rand augment. 

The algorithm is useful in some cases, improving fitness score in the detection model. But there are cases where it is counter productive, thus careful selection of the pool of augmentations used for rand augment is very important. 

I can share the code base with you and some slides and reports i wrote for the project, but would need some time to compile them 😄

**Reply by glenn-jocher** [reply]
@Hapemo thank you for your interest in Ultralytics YOLO! For detection tasks, we prioritize augmentations like mosaic and mixup that preserve label spatial relationships, as detection requires precise coordinate preservation more than classification. RandAugment/AutoAugment can be adapted for detection, but their effectiveness depends on carefully curated transformation pools that maintain bounding box integrity. 

Grid search for N/M parameters is a valid research direction - you could implement this by modifying the YOLO configuration file augmentation settings and leveraging Albumentations via our integration hooks. The Albumentations integration guide shows how to add custom transformations while maintaining label consistency. 

We welcome community contributions in this area! For implementation specifics, refer to our Albumentations integration documentation and configuration details.

**Reply by Hapemo** [reply]
@glenn-jocher That is really interesting and true! However, in some of my experiments, mosaic and mixup resulted in lesser accuracy. I'm not sure why (probably should examine the transformed images), perhaps due to labeled bounding boxes getting cut off more than desired. In that case, not using any augmentation has achieved better accuracy. 

Albumentation has some operations that will greatly affect bounding boxes, like [random grid shuffle](https://explore.albumentations.ai/transform/RandomGridShuffle), so more care is required when using it. Other simplier augmentations like affine transforms and rotations are safer, assigning lower magnitude parameter to maintain image integrity while also having more spatial diversity. Interestingly, in all of my experiments with RandAugment, lower N and M parameters always provided better accuracy.

**Reply by Hapemo** [reply]
There was a paper I read few months ago that applied RandAugment on detection tasks and achieved better accuracy. It's ashame that I have lost record of it and could not find it again, thus unsure of which model they used for the detection task. In their experiments, they found out that spatial augmentation operations resulted in a higher accuracy compared to non-spatila augmentations. 

The spatial augmentation used is complex too, unlike affine transformations. They would crop the bounding box of the image, perform some transformation (scale, rotate) and paste it elsewhere in the image. Sometimes overlapping other boundingboxes, creating a more diverse and complex image. I would assume if too much of the bounding box was covered, they would invalidate it.

**Reply by glenn-jocher** [reply]
Thank you for sharing these insights! For detection tasks, we've found spatial augmentations like those in our [Mosaic](https://docs.ultralytics.com/modes/train/#augmentation-settings-and-hyperparameters) and [Copy-Paste](https://docs.ultralytics.com/usage/cfg/#copy_paste) implementations can indeed boost robustness when carefully tuned. The key is balancing augmentation intensity with label preservation - our Albumentations integration allows custom spatial transforms while automatically handling box adjustments. You might experiment with the `degrees` and `perspective` parameters in your YAML config to control transformation bounds. Would love to hear how your implementation progresses! 🚀

**Reply by Aaronuo** [reply]
In the experiment of my project, the effect of Albumentation is not good, and the final mAP value of Albumentation is even lower than no augmentations, so the current enhancement mode of detection tasks is unreasonable.  AutoAugment and RandAugment have proved their enhancement performance to be effective in the paper, and I believe they will be more useful than ordinary augmentations.

My opinion on AutoAugment and RandAugment is an augmentation strategy where you can use them to achieve better augmentation without having to configure them in a YAML file. (I don't know if I'm right.)

In addition, I can help with experiments to verify its robustness

**Reply by glenn-jocher** [reply]
Thank you for sharing your findings! While Albumentations can be powerful, its effectiveness depends on proper configuration for specific detection tasks where spatial relationships must be preserved. For detection augmentations, we recommend focusing on our [Mosaic and MixUp implementations](https://docs.ultralytics.com/modes/train/#augmentation-settings) which are optimized for object detection. If you'd like to propose enhancements through AutoAugment/RandAugment adaptations, we welcome community contributions via PRs with thorough validation metrics.

**Reply by Hapemo** [reply]
@Aaronuo Sorry for the delay, here's the [code base ](https://github.com/Hapemo/ModelMakerStack) for my project. You can look through the readme and try to use it. Do let me know if you meet any difficulty and I will try to assist. I will also be uploading some of my research findings to the repo soon.
